lib/platform/linux-uart.c

```diff
@@ -0,0 +1,437 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <termios.h>
+#include "../crc8.h"
+
+const char *dbginfo_off = "pscdbg 0 all\r";
+const char *echo_off = "echo 0\r";
+const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(50)
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	int idx;
+	int rty_cnt = 0;
+	char rtn[1024];
+	int cnt = 0;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = write(udev->fd, str, strlen(str));
+	if (ret != strlen(str))
+		return -1;
+
+cli_rtn_retry:
```

Comment:
replace end of response from device check by loop instead of goto

Reply:
Keep the goto for crc error retry.

Reply:
I think you'll find once you refactor your code as I ask you can make a loop out of that retry too.

You should never use gotos for loops. Refactor your code as I've asked so you don't need the spaghetti gotos. It's already starting to look more readable but could use more.

Reply:
The first goto is already replaced with loop, cause it is a normal case, wait for the response from FW till finished, once finished, break the loop.
But the second goto is for  the crc error case, instead a normal case loop.

Actually, I find replace the second goto with loop, will make the code unreadable.

Reply:
You're wrong. Only because it's currently a mess does it look better with the goto. Goto's should simply *never* be used where a loop primitive can be used.